src/breaker/SamplingBreaker.ts

```diff
@@ -0,0 +1,172 @@
+import { CircuitState } from '../CircuitBreakerPolicy';
+import { IBreaker } from './Breaker';
+
+interface IWindow {
+  startedAt: number;
+  failures: number;
+  successes: number;
+}
+
+export interface ISamplingBreakerOptions {
+  /**
+   * Percentage (from 0 to 1) of requests that need to fail before we'll
+   * open the circuit.
+   */
+  threshold: number;
+
+  /**
+   * Length of time over which to sample.
+   */
+  duration: number;
+
+  /**
+   * Minimum number of RPS needed to be able to (potentially) open the circuit.
+   * Useful to avoid unnecessarily tripping under low load.
+   */
+  minimumRps?: number;
+}
+
+/**
+ * SamplingBreaker breaks if more than `threshold` percentage of calls over the
+ * last `samplingDuration`, so long as there's at least `minimumRps` (to avoid
+ * closing unnecessarily under low RPS).
+ */
+export class SamplingBreaker implements IBreaker {
+  private readonly threshold: number;
+  private readonly minimumRpms: number;
+  private readonly duration: number;
+  private readonly windowSize: number;
+
+  private windows: IWindow[] = [];
+  private currentWindow = 0;
+  private currentFailures = 0;
+  private currentSuccesses = 0;
+
+  /**
+   * Gets a snapshot of internal breaker state. Used for tests.
+   */
+  public get state() {
+    return {
+      threshold: this.threshold,
+      minimumRpms: this.minimumRpms,
+      duration: this.duration,
+      windowSize: this.windowSize,
+      windows: this.windows.map(w => ({ ...w })),
+      currentWindow: this.currentWindow,
+      currentFailures: this.currentFailures,
+      currentSuccesses: this.currentSuccesses,
+    };
+  }
+
+  constructor({ threshold, duration: samplingDuration, minimumRps }: ISamplingBreakerOptions) {
+    if (threshold <= 0 || threshold >= 1) {
+      throw new Error(`SamplingBreaker threshold should be between (0, 1), got ${threshold}`);
+    }
+
+    this.threshold = threshold;
+
+    // at least 5 windows, max 1 second each:
+    const windowCount = Math.max(5, Math.ceil(samplingDuration / 1000));
+    for (let i = 0; i < windowCount; i++) {
+      this.windows.push({ startedAt: 0, failures: 0, successes: 0 });
+    }
+
+    this.windowSize = Math.round(samplingDuration / windowCount);
+    this.duration = this.windowSize * windowCount;
+
+    if (minimumRps) {
+      this.minimumRpms = minimumRps / 1000;
```

Comment:
This strikes me as a little odd, you might want to keep the external and internal unit equal.
In js it's very common to use ms as the principal unit for time (see Date.now, setTimeout etc.)

Reply:
Yea. We mainly do this conversion here to avoid doing extra conversion work in the 'hot path'. But "rps" in the common, colloquial term that I want to expose; it's also easier to reason about for humans, thinking their API gets 10 rps instead of 0.01 rpms